pkg/scale/encode.go

```diff
@@ -0,0 +1,420 @@
+package scale
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+func Marshal(v interface{}) (b []byte, err error) {
+	es := encodeState{
+		fieldScaleIndicesCache: cache,
+	}
+	err = es.marshal(v)
+	if err != nil {
+		return
+	}
+	b = es.Bytes()
+	return
+}
+
+type encodeState struct {
+	bytes.Buffer
+	*fieldScaleIndicesCache
+}
+
+func (es *encodeState) marshal(in interface{}) (err error) {
+	switch in := in.(type) {
+	case int:
+		err = es.encodeUint(uint(in))
+	case uint:
+		err = es.encodeUint(in)
+	case int8, uint8, int16, uint16, int32, uint32, int64, uint64:
+		err = es.encodeFixedWidthInt(in)
+	case *big.Int:
+		err = es.encodeBigInt(in)
+	case *Uint128:
+		err = es.encodeUint128(in)
+	case []byte:
+		err = es.encodeBytes(in)
+	case string:
+		err = es.encodeBytes([]byte(in))
+	case bool:
+		err = es.encodeBool(in)
+	default:
+		switch reflect.TypeOf(in).Kind() {
+		case reflect.Ptr:
+			// Assuming that anything that is a pointer is an Option to capture {nil, T}
+			elem := reflect.ValueOf(in).Elem()
+			switch elem.IsValid() {
```

Comment:
why `IsValid` instead of `IsNil`?

Reply:
>IsValid reports whether v represents a value. It returns false if v is the zero Value. If IsValid returns false, all other methods except String panic. Most functions and methods never return an invalid Value. If one does, its documentation states the conditions explicitly.

`IsNil` will panic if it isn't valid. 🤷

Reply:
hm interesting, that seems weird on part of the reflect library but okay